pkg/kubelet/checkpointmanager/checkpoint_manager_test.go

```diff
@@ -0,0 +1,245 @@
+}
+
+func (cp *CheckpointData) VerifyChecksum() error {
+	return cp.Checksum.Verify(*cp.Data)
+}
+
+func (cp *CheckpointData) GetData() ([]*PortMapping, bool) {
+	return cp.Data.PortMappings, cp.Data.HostNetwork
+}
+
+type checkpointDataV2 struct {
+	Version  string
+	Name     string
+	Data     *CheckpointDataV2
+	Checksum checksum.Checksum
+}
+
+func newFakeCheckpointV2(name string, portMappings []*PortMapping, hostNetwork bool) FakeCheckpoint {
+	return &checkpointDataV2{
+		Version: "v2",
+		Name:    name,
+		Data: &CheckpointDataV2{
+			PortMappings: portMappings,
+			HostNetwork:  hostNetwork,
+		},
+	}
+}
+
+func newFakeCheckpointRemoteV1(name string, portMappings []*v1.PortMapping, hostNetwork bool) Checkpoint {
+	return &v1.CheckpointData{
+		Version: "v1",
+		Name:    name,
+		Data: &v1.Data{
+			PortMappings: portMappings,
+			HostNetwork:  hostNetwork,
+		},
+	}
+}
+
+func (cp *checkpointDataV2) MarshalCheckpoint() ([]byte, error) {
+	cp.Checksum = checksum.New(*cp.Data)
+	return json.Marshal(*cp)
+}
+
+func (cp *checkpointDataV2) UnmarshalCheckpoint(blob []byte) error {
+	return json.Unmarshal(blob, cp)
+}
+
+func (cp *checkpointDataV2) VerifyChecksum() error {
+	return cp.Checksum.Verify(*cp.Data)
+}
+
+func (cp *checkpointDataV2) GetData() ([]*PortMapping, bool) {
+	return cp.Data.PortMappings, cp.Data.HostNetwork
+}
+
+func newTestCheckpointManager() CheckpointManager {
+	return &impl{store: testStore}
+}
+
+func TestCheckpointManager(t *testing.T) {
+	var err error
+	testStore = utilstore.NewMemStore()
+	manager := newTestCheckpointManager()
+	port80 := int32(80)
+	port443 := int32(443)
+	proto := protocol("tcp")
+
+	portMappings := []*PortMapping{
+		{
+			&proto,
+			&port80,
+			&port80,
+		},
+		{
+			&proto,
+			&port443,
+			&port443,
+		},
+	}
+	checkpoint1 := newFakeCheckpointV1("check1", portMappings, true)
+
+	checkpoints := []struct {
+		checkpointKey     string
+		checkpoint        FakeCheckpoint
+		expectHostNetwork bool
+	}{
+		{
+			"key1",
+			checkpoint1,
+			true,
+		},
+		{
+			"key2",
+			newFakeCheckpointV1("check2", nil, false),
+			false,
+		},
+	}
+
+	for _, tc := range checkpoints {
+		// Test CreateCheckpoints
+		err = manager.CreateCheckpoint(tc.checkpointKey, tc.checkpoint)
+		assert.NoError(t, err)
+
+		// Test GetCheckpoints
+		checkpointOut := newFakeCheckpointV1("", nil, false)
+		err := manager.GetCheckpoint(tc.checkpointKey, checkpointOut)
+		assert.NoError(t, err)
+		actualPortMappings, actualHostNetwork := checkpointOut.GetData()
+		expPortMappings, expHostNetwork := tc.checkpoint.GetData()
+		assert.Equal(t, actualPortMappings, expPortMappings)
+		assert.Equal(t, actualHostNetwork, expHostNetwork)
+	}
+	// Test it fails if tried to read V1 structure into V2, a different structure from the structure which is checkpointed
+	checkpointV2 := newFakeCheckpointV2("", nil, false)
+	err = manager.GetCheckpoint("key1", checkpointV2)
+	assert.EqualError(t, err, "checkpoint is corrupted")
```

Comment:
I spent some time looking into this today. I think the reason checksum checking fails here is because the hashutil.DeepHashObject() function used in getChecksum() uses spew printer that also prints out the field type. Here are what it generates from spew printer.Fprintf:

for checkpoint1:
(checkpointmanager.Data){PortMappings:([]*checkpointmanager.PortMapping)[<*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)80 HostPort:(*int32)80} <*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)443 HostPort:(*int32)443}] HostNetwork:(bool)true}

for checkpoint2:
(checkpointmanager.CheckpointDataV2){PortMappings:([]*checkpointmanager.PortMapping)[<*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)80 HostPort:(*int32)80} <*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)443 HostPort:(*int32)443}] HostNetwork:(bool)true V2Field:(string)}

Basically, this means that checksum checking would always fail whenever we add new fields to the checkpointed data structure, even though json Unmarshal would pass. Looking at the change, I did notice that this is also the current behavior in docker_checkpoint.go, but @yujuhong @Random-Liu and @dchen1107 can you verify that this is indeed the desired behavior for kubelet?

Reply:
Good point! Today we shouldn't run into any issue since we are doing a bundled upgrade on a node, hence there wouldn't have a new field be introduced to the checkpointed data structure when kubelet is restarted without tearing-down-then-recreation. But once we introduced in-place-upgrade, we might encounter the issue. @yujuhong is it possible to change such behavior in docker-shim's checkpoints? Or we want to keep that to prevent arbitrary checkpointed data structure changes?

Reply:
The purpose of the checksum is to detect corruption. I *think* with dockershim's checkpoint, you're supposed to version the checkpoint struct, and adding/changing a field requires a version bump. Alternatively, you can also hash the raw content of the file, but that requires saving the checksum as a separate file. +@freehan (the author of dockershim's checkpointing) in case I'm wrong.

Reply:
@jiayingz Done, PTAL!!